packages/loader/container-loader/src/container.ts

```diff
@@ -674,17 +675,65 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         this.resumeInternal();
     }
 
+    /**
+     * Waits until container connects to delta storage and gets up-to-date
+     * Useful when resolving URIs and hitting 404, due to container being loaded from (stale) snapshot and not being
+     * up to date. Host may chose to wait in such case and retry resolving URI.
+     * @returns true: container is up to date, it processed all the ops that were know at the time of first connection
+     *          false: storage does not provide indication of how far the client is.
+     */
+    public async waitCatchUp() {
+        return new Promise<boolean>((accept) => {
+            const waitForOps = () => {
+                assert(this._connectionState !== ConnectionState.Disconnected);
+                if (!this.hasCheckpointSequenceNumber) {
+                    accept(false);
+                    return;
+                }
+                const connectionOpSeqNumber = this.deltaManager.lastKnownSeqNumber;
+                if (this.deltaManager.lastSequenceNumber === connectionOpSeqNumber) {
+                    accept(true);
+                    return;
+                }
+                const callbackOps = (message) => {
+                    if (connectionOpSeqNumber <= message.sequenceNumber) {
+                        accept(true);
```

Comment:
Since this is "beforeOpProcessing", this is resolving before the final op gets processed, right?  I think you're taking an implicit dependency here on the processing being synchronous so it completes before anyone awaiting this promise proceeds?

Reply:
Yep. I thought about it too :) I copied code from container runtime (OpPerfTelemetry class). I'll follow up, I think we do not have "after" event - need to add it here.

Reply:
I'm fine with an after event too, but also maybe consider if this logic would be more at home in `DeltaManager` -- basically all of the logic is not dealing with `Container`'s members, but rather `DeltaManager`'s.

Reply:
Thanks, I'll take a look. It's funny that we re doing full circle here - most of the logic of OpPerfTelemetry was part of DeltaManager, and as I discussed it with Kurt, we agreed to move it out to keep this layer simple and fixed. This case is slightly different as host needs it, so it has to be at that layer. The only reason I put it in Container is that it need to start somewhere, but I feel like it's better to be just flat function outside of Container/DeltaManager - I actually started there, but then needed to add one member variable, so that now ties it to either Container or DeltaManager

Reply:
In some of the stuff I've been prototyping I've been looking at just emitting a simple event for when a particular queue thinks it's caught up which seems pretty ok so far (and seems like something broadly useful for consumers of a processing queue to observe).

Reply:
This is less about queue, more about knowledge of what is still not in the queue and needs to gets there and be processed.

Reply:
Added "op" event and did some cleanup in this space. Please take a look. Removed events are not used anywhere in either repo, and I assume we should push people to using deltaManager directly vs. keep proxying everything to every layer (though may use adapters / proxies to hide certain capabilities at different levels)